### ISSUE TITLE
Inheritance - exec with --controls filters out required dependent controls

### ISSUE THREAD

**Issue by kevin-j-smith** [inspec/inspec#4318]
# Describe the problem
I am trying to nest my controls for better orchestration of what is ran across development pipelines.  I also want to nest my controls into groups.   Take for instance
```
inheritance
|-- controls
   |-- dev.rb
|-- inspec.yml
|-- profile_group_1
   |-- controls
      |-- c1.rb
   |-- inspec.yml
```

```ruby
# dev.rb
require_controls 'group_1' do
  control 'c1'
end
```

```ruby
# inspec.yml
name: inheritance
version: 0.1.0
depends:
  - name: group_1
     path: ./profile_group_1
```

```ruby
# c1.rb
control 'c1' do
  title 'hello'
  desc 'world'
  impact 0.5
  describe 'HELLO' do
    it { should cmp 'HELLO' }
  end
end
```

When I run: 
```bash
inspec exec --no-create-lockfile -- .
```

I get what I expect to get, though I would accept it if No tests executed was not given.
```bash
Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

  ✔  c1: hello
     ✔  HELLO should cmp == "HELLO"


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 1 successful, 0 failures, 0 skipped
```

Now if I expand my profile with stage.rb
inheritance
|-- controls
   |-- dev.rb
   |-- stage.rb
|-- inspec.yml
|-- profile_group_1
   |-- controls
      |-- c1.rb
   |-- inspec.yml

```ruby
# stage.rb
require_controls 'group_1' do
  control 'c1'
end
```

And run the same command:
```bash
Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

  ✔  c1: hello
     ✔  HELLO should cmp == "HELLO"
     ✔  HELLO should cmp == "HELLO"


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 2 successful, 0 failures, 0 skipped
```

This is only somewhat expected.  And acceptable, that c1 is ran twice but with two results.  I could go either way: 
1. One result
2. Multiple results but the results are nested under the parent's results.   This would be better especially if one control uses the child control with no enhancements and the other uses:
  ```ruby
  control 'c1' do
    impact 0.0
  end
  ```

Now the fun.  I don't want to run stage in the dev environment (and I don't want to have to pass around attributes in all controls to skip when necessary or maintain some comparison list of what is expected to be skipped).   So I run: 
```bash
inspec exec --no-create-lockfile --controls dev -- .

Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

     No tests executed.

Test Summary: 0 successful, 0 failures, 0 skipped
```

This is unexpected.   I was expecting the controls flag only to manage the parents controls and not the depended-upon controls.  After much testing and code review I noticed that c1 was being filtered out.  So I can run:
```bash
inspec exec --no-create-lockfile --controls dev /c/ -- .
Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

  ✔  c1: hello
     ✔  HELLO should cmp == "HELLO"
     ✔  HELLO should cmp == "HELLO"


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 2 successful, 0 failures, 0 skipped
```

Again, not what I was expecting.  Since I am only running dev and anything the 'c' in it why is c1 being ran twice?  Maybe the directory name 'controls' was being referenced in the matcher.  So:
```bash
inspec exec --no-create-lockfile --controls dev /c1/ -- .

Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

  ✔  c1: hello
     ✔  HELLO should cmp == "HELLO"
     ✔  HELLO should cmp == "HELLO"


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 2 successful, 0 failures, 0 skipped
```

Argh, same results.  Ok let's have some more fun.  Let's say that dev requires c1 and stage requires c2.  There is no overlap, yet:
inheritance
|-- controls
   |-- dev.rb
   |-- stage.rb
|-- inspec.yml
|-- profile_group_1
   |-- controls
      |-- c1.rb
      |-- c2.rb
   |-- inspec.yml

```ruby
# c2.rb
control 'c2' do
  title 'hello'
  desc 'world'
  impact 0.5
  describe 'WORLD' do
    it { should cmp 'WORLD' }
  end
end
```

```ruby
# stage.rb
require_controls 'group_1' do
  control 'c2'
end
```

Another unexpected result:
```bash
inspec exec --no-create-lockfile --controls dev /c/ -- .

/inspec-4.6.4/lib/inspec/dsl.rb:84:in `block in filter_included_controls': undefined method `+' for nil:NilClass (NoMethodError)
/inspec-4.6.4/lib/inspec/dsl.rb:82:in `each'
/inspec-4.6.4/lib/inspec/dsl.rb:82:in `filter_included_controls'
/inspec-4.6.4/lib/inspec/dsl.rb:71:in `load_spec_files_for_profile'
/inspec-4.6.4/lib/inspec/dsl.rb:8:in `require_controls'
/inspec-4.6.4/lib/inspec/profile_context.rb:160:in `instance_eval'
/inspec-4.6.4/lib/inspec/profile_context.rb:160:in `load_with_context'
/inspec-4.6.4/lib/inspec/profile_context.rb:144:in `load_control_file'
/inspec-4.6.4/lib/inspec/profile.rb:195:in `block in collect_tests'
/inspec-4.6.4/lib/inspec/profile.rb:192:in `each'
/inspec-4.6.4/lib/inspec/profile.rb:192:in `collect_tests'
/inspec-4.6.4/lib/inspec/runner.rb:110:in `block in load'
/inspec-4.6.4/lib/inspec/runner.rb:93:in `each'
/inspec-4.6.4/lib/inspec/runner.rb:93:in `load'
/inspec-4.6.4/lib/inspec/runner.rb:121:in `run'
/inspec-4.6.4/lib/inspec/cli.rb:297:in `exec'
/thor-0.20.3/lib/thor/command.rb:27:in `run'
/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
/thor-0.20.3/lib/thor.rb:387:in `dispatch'
/thor-0.20.3/lib/thor/base.rb:466:in `start'
/inspec-bin-4.6.4/bin/inspec:11:in `<top (required)>'
```

After adding some basic debug output, I can see:
```ruby
# lib/inspec/dsl.rb
# Line: 77-92
  def self.filter_included_controls(context, profile, &block)
    puts "HELLO: #{context}\n#{profile}\n#{block}"
    mock = Inspec::Backend.create(Inspec::Config.mock)
    include_ctx = Inspec::ProfileContext.for_profile(profile, mock)
    include_ctx.load(block) if block_given?
    # remove all rules that were not registered
    puts "CONTEXT.all_rules: #{context.all_rules}"
    context.all_rules.each do |r|
      id = Inspec::Rule.rule_id(r)
      puts "WORLD: #{r}\n#{id}"
      fid = Inspec::Rule.profile_id(r) + "/" + id
      unless include_ctx.rules[id] || include_ctx.rules[fid]
        context.remove_rule(fid)
      end
    end
  end
```

```bash
inspec exec --no-create-lockfile --controls dev /c/ -- .

HELLO: #<Inspec::ProfileContext:0x00007fb08b962150>
Inspec::Profile<hello>
#<Proc:0x00007fb04b3408e8@../inheritance/controls/dev.rb:2>
CONTEXT.all_rules: [#<#<Class:0x00007fb07b3033f0>:0x00007fb07b249ce8 @impact=0.5, @title="hello", @descriptions={:default=>"world"}, @refs=[], @tags={}, @__code=nil, @__block=#<Proc:0x00007fb07b24a170@inheritance/profile_group_1/controls/c1.rb", @__group_title=nil>]
WORLD: c2
c2
WORLD: c1
c1
HELLO: #<Inspec::ProfileContext:0x00007fb08b962150>
Inspec::Profile<hello>
#<Proc:0x00007fb04b291758@../inheritance/controls/stage.rb:2>
CONTEXT.all_rules: [nil, #<#<Class:0x00007fb07b3033f0>:0x00007fb07b208bf8 @impact=0.5, @title="hello", @descriptions={:default=>"world"}, @refs=[{:ref=>[]}], @tags={}, @__code=nil, @__block=#<Proc:0x00007fb07b208c48@inheritance/profile_group_1/controls/c1.rb", @__group_title=nil>]
WORLD: 
bundler: failed to load command: inspec-bin/bin/inspec (inspec-bin/bin/inspec)
NoMethodError: undefined method `+' for nil:NilClass
  inspec/lib/inspec/dsl.rb:87:in `block in filter_included_controls'
  inspec/lib/inspec/dsl.rb:84:in `each'
  inspec/lib/inspec/dsl.rb:84:in `filter_included_controls'
  inspec/lib/inspec/dsl.rb:71:in `load_spec_files_for_profile'
  inspec/lib/inspec/dsl.rb:8:in `require_controls'
  ../inheritance/controls/stage.rb:2:in `load_with_context'
  inspec/lib/inspec/profile_context.rb:161:in `instance_eval'
  inspec/lib/inspec/profile_context.rb:161:in `load_with_context'
  inspec/lib/inspec/profile_context.rb:145:in `load_control_file'
  inspec/lib/inspec/profile.rb:198:in `block in collect_tests'
  inspec/lib/inspec/profile.rb:194:in `each'
  inspec/lib/inspec/profile.rb:194:in `collect_tests'
  inspec/lib/inspec/runner.rb:110:in `block in load'
  inspec/lib/inspec/runner.rb:93:in `each'
  inspec/lib/inspec/runner.rb:93:in `load'
  inspec/lib/inspec/runner.rb:121:in `run'
  inspec/lib/inspec/cli.rb:297:in `exec'
  .rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  .rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  .rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  .rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
  inspec/lib/inspec/base_cli.rb:33:in `start'
  inspec-bin/bin/inspec:11:in `<top (required)>'
```

Same happens with the command: 'inspec exec --no-create-lockfile --controls dev /c/ -- .'

Now if I remove stage.rb, things work magically again:
```bash
inspec exec --no-create-lockfile -- .
Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

  ✔  c1: hello
     ✔  HELLO should cmp == "HELLO"


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 1 successful, 0 failures, 0 skipped
```

Ok last scenario, all dependent controls seem to be loaded and evaluated even if they are not going to be ran.   I specifically use attribute here for the warning message to see what gets loaded.   c2 should not get loaded.

```ruby
# c1.rb
hello = attribute('hello', default: "HELLO", description: 'nice')

control 'c1' do
  title 'hello'
  desc 'world'
  impact 0.5
  describe hello do
    it { should cmp hello }
  end
end
```

```ruby
# c2.rb

world = attribute('world', default: 'WORLD', description: 'groovy')

control 'c2' do
  title 'hello'
  desc 'world'
  impact 0.5
  describe world do 
    it { should cmp world }
  end
end
```

```bash
inspec exec --no-create-lockfile --controls dev /c/ -- .
[2019-07-18T11:52:46-05:00] WARN: DEPRECATION: The 'default' option for attributes is being replaced by 'value' - please use it instead. attribute name: 'Inspec::Input' (used at /Users/ks066664/git/kevin-j-smith/inheritance/profile_group_1/controls/c2.rb:1)
[2019-07-18T11:52:46-05:00] WARN: DEPRECATION: The 'default' option for attributes is being replaced by 'value' - please use it instead. attribute name: 'Inspec::Input' (used at /Users/ks066664/git/kevin-j-smith/inheritance/profile_group_1/controls/c1.rb:1)

Profile: itance (inher)
Version: 0.1.0
Target:  local://

     No tests executed.

Profile: world (hello)
Version: 0.1.0
Target:  local://

  ✔  c1: hello
     ✔  HELLO should cmp == "HELLO"


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 1 successful, 0 failures, 0 skipped

```

Here are the issues:
1. Inheritance does not seem to work when multiple controls are relying on dependent profiles (at least in this case the same dependent profile)
1. Inheritance loads all controls in the dependent profile then filters them.   In some controls there is logic that is ran and this just does unnecessary work.  When the --controls flag is set, only the parents controls should be filtered.  All other filtering should come from the include_controls and require_controls statements

## Possible Solution
I am not familiar with Inspecs profile/control loading sequence to be helpful here.  I would assume that the --controls flag allows for filtering controls and this should be done before loading controls or running any logic in them.   As to multiple controls relying on the same depended-upon profile, I would hope that I can use --controls and not have to supply a regex for the depended-upon controls to get it working.  

Else, I could be missing some major Inspec feature that already does what I am describing.

**Comment by zenspider** [comment]
Please split this up into multiple issues.

**Comment by Vasu1105** [comment]
1) The error `undefined method `+' for nil:NilClass (NoMethodError)` should get resolved.
2) Does --controls option should work with dependent controls ? as we already have `skip_control` option to be used with require_controls or include_controls? 
3) Can we filter on the profile name using option --controls like using is trying to filter on `dev` ? As per my understanding we can't as --controls options is mean to be filter the controls with the given name.
For e.g. I have `dev.rb` profile under controls which has following entries
```
control 'test1' do
  title 'hello'
  desc 'world'
  impact 0.5
  describe 'HELLO' do
    it { should cmp 'HELLO' }
  end
end

control 'test2' do
  title 'test'
  desc 'world'
  impact 0.5
  describe 'HELLO' do
    it { should cmp 'HELLO' }
  end
end
```
If I want to filter only test1 control then I can run something like `inspec exec my-test-profiles/inheritance/ --no-create-lockfile --controls /test1/` or `inspec exec my-test-profiles/inheritance/ --no-create-lockfile --controls test1` 
but I think we can;t filter on dev like `inspec exec my-test-profiles/inheritance/ --no-create-lockfile --controls dev`

4) What should be expected result when we use --controls option with depedent profiles using require_controls or include_controls ?

**Comment by Vasu1105** [comment]
1) We do not support the filtering using --controls option by profile controls file name. The --control option works with control name.
2) To achieve the different environment scenario you can maintain two different profiles. Which can be run individually on those environment.
3) Also we have fixed the way of --control option getting evaluated in this PR https://github.com/inspec/inspec/pull/5434.
4) We have fixed undefined method error which is listed in the bunch of scenarios here is PR which is merged https://github.com/inspec/inspec/pull/5487
I am closing this issue for now as many of the scenarios can be achieved or can be worked out in different way.
